spark3/src/main/java/org/apache/iceberg/actions/Spark3CreateAction.java

```diff
@@ -0,0 +1,183 @@
+
+    this.spark = spark;
+    this.sourceCatalog = checkSourceCatalog(sourceCatalog);
+    this.sourceTableName = sourceTableName;
+    this.destCatalog = checkDestinationCatalog(destCatalog);
+    this.destTableName = destTableName;
+
+    try {
+      String sourceString = String.join(".", sourceTableName.namespace()) + "." + sourceTableName.name();
+      sourcePartitionSpec = SparkSchemaUtil.specForTable(spark, sourceString);
+    } catch (AnalysisException e) {
+      throw new IllegalArgumentException("Cannot determining partitioning of " + sourceTableName.toString(), e);
+    }
+
+    try {
+      this.sourceTable = (V1Table) ((TableCatalog) sourceCatalog).loadTable(sourceTableName);
+      this.sourceCatalogTable = sourceTable.v1Table();
+    } catch (NoSuchTableException e) {
+      throw new IllegalArgumentException(String.format("Cannot not find source table %s", sourceTableName), e);
+    } catch (ClassCastException e) {
+      throw new IllegalArgumentException(String.format("Cannot use a non-v1 table %s as a source", sourceTableName), e);
+    }
+    validateSourceTable(sourceCatalogTable);
+
+    this.sourceTableLocation = CatalogUtils.URIToString(sourceCatalogTable.storage().locationUri().get());
+  }
+
+  @Override
+  public CreateAction setAll(Map<String, String> properties) {
+    this.additionalProperties.putAll(properties);
+    return this;
+  }
+
+  @Override
+  public CreateAction set(String key, String value) {
+    this.additionalProperties.put(key, value);
+    return this;
+  }
+
+  protected SparkSession spark() {
+    return spark;
+  }
+
+  protected String sourceTableLocation() {
+    return sourceTableLocation;
+  }
+
+  protected CatalogTable v1SourceTable() {
+    return sourceCatalogTable;
+  }
+
+  protected CatalogPlugin sourceCatalog() {
+    return sourceCatalog;
+  }
+
+  protected Identifier sourceTableName() {
+    return sourceTableName;
+  }
+
+  protected Transform[] sourcePartitionSpec() {
+    return sourceTable.partitioning();
+  }
+
+  protected StagingTableCatalog destCatalog() {
+    return destCatalog;
+  }
+
+  protected Identifier destTableName() {
+    return destTableName;
+  }
+
+  protected Map<String, String> additionalProperties() {
+    return additionalProperties;
+  }
+
+  private static void validateSourceTable(CatalogTable sourceTable) {
+    String sourceTableProvider = sourceTable.provider().get().toLowerCase(Locale.ROOT);
+
+    if (!ALLOWED_SOURCES.contains(sourceTableProvider)) {
+      throw new IllegalArgumentException(
+          String.format("Cannot create an Iceberg table from source provider: %s", sourceTableProvider));
+    }
+
+    if (sourceTable.storage().locationUri().isEmpty()) {
+      throw new IllegalArgumentException("Cannot create an Iceberg table from a source without an explicit location");
+    }
+  }
+
+  protected static Map<String, String> tableLocationProperties(String tableLocation) {
```

Comment:
Why do we have this method in parent class? Isn't the location config specific to the operation type? We have to set the data location to the table location only in MIGRATE.

Reply:
And with Snapshot with a manually specified location

Reply:
I am not sure I follow. Why do we have to force a specific layout if the user specifies a location in `SNAPSHOT`?

It should be OK to specify a custom metadata and data location in the SNAPSHOT command and we are forcing the layout here. In my view, this should only apply to `MIGRATE` where we must ensure the old layout.

Reply:
This only applies if the user also uses .withLocation in SnapshotAction, if "destLocation" is not set then these properties are not set.

Reply:
Correct but why? Why cannot we set a location and the data and metadata folder in SNAPSHOT?

Reply:
I don't think this is needed either way.

In our MIGRATE command, we set the metadata location so that we can create the Iceberg table and then rename it into place. Because it doesn't have the correct name to begin with, its location would be different so we set the metadata location to ensure the metadata is created under the correct location. I think we also didn't support custom table locations at the time I wrote it.

Now, tables support custom table locations in create and this is going to use a staged table created to create in place. I think as long as the table is created from the exiting table's location for MIGRATE, or created with the requested location for SNAPSHOT, everything should work fine without the additional table properties.

Reply:
@rdblue, what about keeping the old layout after migration? I agree nothing is required for SNAPSHOT but I think we should keep the existing layout of data after migration (i.e. data location for new data should point to the root table location).

Reply:
I would recommend against that because it modifies the state of the partitions in the old table. We use the `data` prefix so that the old table can be kept unmodified in case the operation needs to be rolled back, even after a write. Writing data into the old folders makes it live in the old table.

Reply:
Well, it probably depends on the use case. Some people may actually need a way to move back without loosing writes after all snapshots are expired. Also, it is a bit surprising for folks less familiar with Iceberg that the data is written to `data` folder by default.

Reply:
Rolling back the table should be also possible even if we write the data to the same location. We can call rollback and then expire/remove orphan files. That said, I will be okay if we allow extra properties to be set during MIGRATE and document this properly.